Mathlib/Algebra/CharZero/Quotient.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2022 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+
+! This file was ported from Lean 3 source module algebra.char_zero.quotient
+! leanprover-community/mathlib commit d90e4e186f1d18e375dcd4e5b5f6364b01cb3e46
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.GroupTheory.QuotientGroup
+
+/-!
+# Lemmas about quotients in characteristic zero
+-/
+
+
+variable {R : Type _} [DivisionRing R] [CharZero R] {p : R}
+
+namespace AddSubgroup
+
+/-- `z • r` is a multiple of `p` iff `r` is `pk/z` above a multiple of `p`, where `0 ≤ k < |z|`. -/
+theorem zsmul_mem_zmultiples_iff_exists_sub_div {r : R} {z : ℤ} (hz : z ≠ 0) :
+    z • r ∈ AddSubgroup.zmultiples p ↔
+      ∃ k : Fin z.natAbs, r - (k : ℕ) • (p / z : R) ∈ AddSubgroup.zmultiples p := by
+  rw [AddSubgroup.mem_zmultiples_iff]
+  simp_rw [AddSubgroup.mem_zmultiples_iff, div_eq_mul_inv, ← smul_mul_assoc, eq_sub_iff_add_eq]
+  have hz' : (z : R) ≠ 0 := Int.cast_ne_zero.mpr hz
+  conv_rhs => simp (config := { singlePass := true }) only [← (mul_right_injective₀ hz').eq_iff]
+  simp_rw [← zsmul_eq_mul, smul_add, ← mul_smul_comm, zsmul_eq_mul (z : R)⁻¹, mul_inv_cancel hz',
+    mul_one, ← coe_nat_zsmul, smul_smul, ← add_smul]
+  constructor
+  · rintro ⟨k, h⟩
+    simp_rw [← h]
+    refine' ⟨⟨(k % z).toNat, _⟩, k / z, _⟩
+    · rw [← Int.ofNat_lt, Int.toNat_of_nonneg (Int.emod_nonneg _ hz)]
+      exact (Int.emod_lt _ hz).trans_eq (Int.abs_eq_natAbs _)
+    rw [Fin.val_mk, Int.toNat_of_nonneg (Int.emod_nonneg _ hz)]
+    nth_rewrite 3 [← Int.div_add_mod k z]
+    rw [Int.mod_def, ← Int.div_def', Int.emod_def]
+    simp only [add_sub_cancel'_right, zsmul_eq_mul, Int.div_def']
+  · rintro ⟨k, n, h⟩
+    exact ⟨_, h⟩
+#align add_subgroup.zsmul_mem_zmultiples_iff_exists_sub_div
+    AddSubgroup.zsmul_mem_zmultiples_iff_exists_sub_div
+
+theorem nsmul_mem_zmultiples_iff_exists_sub_div {r : R} {n : ℕ} (hn : n ≠ 0) :
+    n • r ∈ AddSubgroup.zmultiples p ↔
+      ∃ k : Fin n, r - (k : ℕ) • (p / n : R) ∈ AddSubgroup.zmultiples p := by
+  rw [← coe_nat_zsmul r, zsmul_mem_zmultiples_iff_exists_sub_div (Int.coe_nat_ne_zero.mpr hn),
+    Int.cast_ofNat]
+  rfl
+#align add_subgroup.nsmul_mem_zmultiples_iff_exists_sub_div
+    AddSubgroup.nsmul_mem_zmultiples_iff_exists_sub_div
+
+end AddSubgroup
+
+namespace quotientAddGroup
+
+theorem zmultiples_zsmul_eq_zsmul_iff {ψ θ : R ⧸ AddSubgroup.zmultiples p} {z : ℤ} (hz : z ≠ 0) :
+    z • ψ = z • θ ↔ ∃ k : Fin z.natAbs, ψ = θ + (k : ℕ) • (p / z : R) := by
+  induction ψ using Quotient.inductionOn'
+  induction θ using Quotient.inductionOn'
+  have : (Quotient.mk'' : R → R ⧸ AddSubgroup.zmultiples p) = coe := rfl
```

Comment:
Currently stuck on how to port the `coe` here.

Reply:
Replacing `coe` by the usual arrow shenanigans works, so `((↑) : R → R ⧸ Zp)`. Together with some mild fixes to some other proofs, that's the file done as far as I can tell!

Reply:
perfect, thanks!